rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,503 @@
+        }
+    }
+
+    unsafe fn thread_static_roots(&self) {
+        let root_array = self.heap.roots.static_roots.as_array();
+        for i in 0..root_array.len() {
+            let object = root_array.get(i).as_obj();
+            debug_assert_eq!(object.tag(), TAG_MUTBOX);
+            debug_assert!((object as usize) < self.heap.limits.base);
+            self.thread_root_mutbox_fields(object as *mut MutBox);
+        }
+    }
+
+    unsafe fn thread_root_mutbox_fields(&self, mutbox: *mut MutBox) {
+        let field_addr = &mut (*mutbox).field;
+        if pointer_to_dynamic_heap(field_addr, self.generation_base()) {
+            self.thread(field_addr);
+        }
+    }
+
+    unsafe fn thread_all_backward_pointers(&mut self) {
+        let mut bitmap_iter = iter_bits();
+        let mut bit = bitmap_iter.next();
+        while bit != BITMAP_ITER_END {
+            let object = (bit * WORD_SIZE) as *mut Obj;
+            self.thread_backward_pointer_fields(object);
+            bit = bitmap_iter.next();
+        }
+    }
+
+    unsafe fn thread_backward_pointer_fields(&mut self, object: *mut Obj) {
+        debug_assert!(object.tag() < TAG_ARRAY_SLICE_MIN);
+        visit_pointer_fields(
+            &mut (),
+            object,
+            object.tag(),
+            self.generation_base(),
+            |_, field_address| {
+                let field_value = *field_address;
+                // Thread if backwards or self pointer
+                if field_value.get_ptr() <= object as usize {
+                    (&self).thread(field_address);
+                }
+            },
+            |_, _, array| array.len(),
+        );
+    }
+
+    // Thread forward pointers in old generation leading to young generation
+    unsafe fn thread_old_generation_pointers(&mut self) {
+        let mut iterator = REMEMBERED_SET.as_ref().unwrap().iterate();
+        while iterator.has_next() {
+            let location = iterator.current().get_raw() as *mut Value;
+            let value = *location;
+            // value in the location may have changed since recording by the write barrer
+            if value.points_to_or_beyond(self.heap.limits.last_free) {
+                debug_assert!((location as usize) >= self.heap.limits.base);
+                self.thread(location);
+            }
+            iterator.next();
+        }
+    }
+
+    unsafe fn move_phase(&mut self) {
+        REMEMBERED_SET = None; // no longer valid when the moving phase starts
+        let mut free = self.heap.limits.base;
+
+        let mut bitmap_iter = iter_bits();
+        if self.strategy == Strategy::Young {
+            free = self.heap.limits.last_free;
+        }
+        let mut bit = bitmap_iter.next();
+        while bit != BITMAP_ITER_END {
+            let old_pointer = (bit * WORD_SIZE) as *mut Obj;
+            let new_pointer = free;
+
+            // Update backwards references to the object's new location and restore object header
```

Comment:
The threading is performed in two phases for the pointers inside heap objects: First, it only threads and unthreads backward pointers. Second, it focusses on forward pointers. Here, it only recovers the backward pointers. If both directions would be combined in one threading phase, it would be difficult to decide whether later visited pointers are already threaded or not.